src/gui/plotWidget.h

```diff
@@ -0,0 +1,105 @@
+ *                                                                            *
+ * LinescaleGUI is free software: you can redistribute it and/or modify       *
+ * it under the terms of the GNU General Public License as published by       *
+ * the Free Software Foundation, either version 3 of the License, or          *
+ * (at your option) any later version.                                        *
+ *                                                                            *
+ * LinescaleGUI is distributed in the hope that it will be useful,            *
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of             *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the               *
+ * GNU General Public License for more details.                               *
+ *                                                                            *
+ * You should have received a copy of the GNU General Public License          *
+ * along with linescaleGUI. If not, see <http://www.gnu.org/licenses/>.       *
+ ******************************************************************************/
+/**
+ * @file plotWidget.h
+ * @authors Gschwind, Weber, Schoch, Niederberger
+ *
+ * @brief Plotting functionality based on QCustomPlot
+ *
+ */
+
+#pragma once
+#ifndef PLOTWIDGET_H_
+#define PLOTWIDGET_H_
+
+// @TODO Better way to disable this warning for MSVC.
+#if _MSC_VER && !__INTEL_COMPILER
+#pragma warning(push)
+#pragma warning(disable : 4127)
+#include <QCustomPlot/qcustomplot.h>
+#pragma warning(pop)
+#else
+#include <QCustomPlot/qcustomplot.h>
+#endif
+
+#include <QTimer>
+#include <QVector>
+#include <QWidget>
+
+/**
+ * @brief A dynamic line graph chart.
+ *
+ * @todo Allow to zoom with panning (touchscreen).
+ * @todo Maybe implement custom range dialog.
+ * @todo Implement saving the graph as image.
+ */
+class Plot : public QWidget {
+   public:
+    /**
+     * @brief Create a new Plot widget.
+     * @param parent The parent widget or `nullptr`.
+     */
+    Plot(QWidget* parent = nullptr);
+
+    /**
+     * @brief Add a single data point.
+     *
+     * Once the data point is added the plot is updated automatically with a maximum
+     * update rate of 60 Hz.
+     *
+     * @param time Horizontal value (time).
+     * @param force Vertical value (force).
+     */
+    void addData(double time, double force);
+
+    /**
+     * @brief Add a new graph to the plot and use it for all new points added.
+     */
+    void beginNewGraph();
+
+   private slots:
+    void selectionChanged();
+    void mousePress();
+    void mouseWheel();
+    void mouseMove();
+    void contextMenuRequest(QPoint pos);
+    void graphClicked(QCPAbstractPlottable* plottable, int dataIndex, QMouseEvent* event);
+
+    void updatePlot();
+    void disableUpdating();
+
+   private:
+    void clearSelection();
+
+   private:
+    QCustomPlot* customPlot;
+    double minValue = 0.0, maxValue = 0.0;
```

Comment:
The plot is extremly slow if it's polluted with sensor noise.
Increase `maxValue` to 2kN to hide this noise

![image](https://user-images.githubusercontent.com/91913978/208219302-ed6382d9-692b-4572-a2ac-83d0c568e792.png)

Later on we should implement some form of a (low-pass) filter to eliminate this issue.
Or simpler, display a movingAverage in form of `t*currentValue + (1-t)*previousValue`

Reply:
> The plot is extremly slow if it's polluted with sensor noise.

What do you mean by slow? As in it's laggy?

You can always zoom out the y-axis to "hide" the noise. I guess a good default would have to do with the accuracy and signal-to-noise ratio of the LineScale. There is also the issue with the used physical unit. So I think we should leave it like this until we tackle the unit problem. 

> Later on we should implement some form of a (low-pass) filter to eliminate this issue.
> Or simpler, display a movingAverage in form of `t*currentValue + (1-t)*previousValue`

Probably not a good idea to show modified data in the plot (per default). The user should be sure that he sees the raw data. If we do this, we should do it as a setting.

Reply:
> What do you mean by slow? As in it's laggy?

The plot becomes much slower when only the noise is displayed. (On 640Hz and 1280Hz)
See attached screencast (e8d81c7 , currTime += 1.0f/1280.0f; and a sampling frequency of 1280Hz)

https://user-images.githubusercontent.com/91913978/208246425-930b149c-a1ab-4fbd-83ab-72378b38368e.mp4

> Probably not a good idea to show modified data in the plot (per default). The user should be sure that he sees the raw data. If we do this, we should do it as a setting.

Good point, we should show the unmodified data in the plot